### PR TITLE
Pythia GPT-NeoX test case

### PR DESCRIPTION
**Pull request by KeitaW**
*Issue #, if available:*
NA

*Description of changes:*

This PR aims to add a new test case, Pythia model training with GPT-NeoX.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+ && ./configure --enable-orterun-prefix-by-default --prefix=$OPEN_MPI_PATH --with-cuda=${CUDA_HOME} --with-slurm --with-pmix=/opt/pmix \
+ && make -j $(nproc) all \
+ && make install \
+ && ldconfig \
+ && cd / \
+ && rm -rf /tmp/openmpi \
+ && ompi_info --parsable --all | grep mpi_built_with_cuda_support:value \
+ # Verify pmix from /opt/pmix/
+ && ldd /opt/amazon/openmpi/lib/openmpi/mca_pmix_ext3x.so | grep '/opt/pmix/lib/libpmix.so.* ' > /opt/amazon/openmpi-pmix.txt
+####################################################################################################
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile:173]: Suggest to remove the pmix and ompi stanzas, as training doesn't seem to use mpi.

Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile]: Thanks @verdimrc , that makes sense. I have updated dockerfile accordingly.

File: 3.test_cases/15.gpt-neox/1.prepare-data.sbatch [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#SBATCH --job-name=pile-dataprep # name of your job
+#SBATCH --output=logs/%x_%j.out # logfile for stdout
+#SBATCH --error=logs/%x_%j.err # logfile for stderr, remove it to merge both outputs
+#SBATCH --exclusive
+
+# default variables for Enroot, if these variables are defined then use them
+: "${APPS_PATH:=/fsx/apps}"
+: "${IMAGE:=$APPS_PATH/gpt-neox.sqsh}"
+: "${FSX_PATH:=/fsx}"
+: "${DATA_PATH:=$FSX_PATH/pile_subset}"  # use pile to download entire dataset
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/1.prepare-data.sbatch:16]: Be specific, use enwiki dump dataset (or something like that).

File: 3.test_cases/15.gpt-neox/1.prepare-data.sbatch [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+: "${MODEL_PATH:=$FSX_PATH/gpt-neox}"    # use pile to download entire dataset
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/1.prepare-data.sbatch:17]: 1. drop pile from comment
2. does comment need "entire"?

File: 3.test_cases/15.gpt-neox/1.prepare-data.sbatch [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+: "${CONTAINER_MOUNT:=$FSX_PATH:$FSX_PATH}"
+
+DATASET=pile_subset                      # use pile to download entire dataset
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/1.prepare-data.sbatch:20]: 1. Be specific with dataset, which is enwiki dump.
2. drop pile from comment
3. Why does the comment mentions "entire" which contradict the "_subset" in the env var?

File: 3.test_cases/15.gpt-neox/2.train.sbatch [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+#SBATCH --exclusive
+set -uxo pipefail
+
+# default variables for Enroot, if these variables are defined then use them
+: "${APPS_PATH:=/fsx/apps}"
+: "${IMAGE:=$APPS_PATH/gpt-neox.sqsh}"
+: "${FSX_PATH:=/fsx}"
+: "${CONTAINER_MOUNT:=$FSX_PATH:$FSX_PATH}"
+## EFA settings
+export FI_EFA_USE_DEVICE_RDMA=1 # use for p4d
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/2.train.sbatch:23]: Remove all unnecessary `FI_EFA_*` env vars for the aws-ofi-nccl plugin in the container.

File: 3.test_cases/15.gpt-neox/README.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+* Docker, [Pyxis](https://github.com/NVIDIA/pyxis) and [Enroot](https://github.com/NVIDIA/enroot) installed.
+* An FSx for Lustre filesystem mounted on `/fsx`.
+
+We recommend that you set up a Slurm cluster using the templates in the architectures [directory](../../1.architectures). You need to set the following environment variables to run this test case:
+
+```bash
+export APPS_PATH=/fsx/apps
+export ENROOT_IMAGE=$APPS_PATH/gpt-neox.sqsh
+export FSX_PATH=/fsx
+export DATA_PATH=$FSX_PATH/pile_subset     # use pile to download entire dataset (see 4. Data preparation)
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/README.md:19]: 1. Be specific, that the dataset is subset of enwiki dump.
2. drop pile from comment.
3. why does the comment mentions "entire" which contradict the "_subset" in the env var?

File: 3.test_cases/15.gpt-neox/README.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+
+
+## 2. Build the container
+
+Before running training jobs, you need to use a build docker container image. [Enroot](https://github.com/NVIDIA/enroot) will be used to turn the image into unprivileged sandbox for Slurm. 
+Below are the steps you need to follow:
+
+
+1. Build the Docker image with the command below in this directory.
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/README.md:33]: This builds xformers from source which requires lots of RAM. Document how user can avoid the build instance hang due to lack of RAM.

**Review by verdimrc — COMMENTED** [review]
Commented.

File: 3.test_cases/15.gpt-neox/1.prepare-data.sbatch (hunk 2 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/1.prepare-data.sbatch:16]: > Be specific, use enwiki dump dataset (or something like that).

Makes sense. Will do that.

> Drop pile from comment.
> Why does the comment mentions "entire" which contradict the "_subset" in the env var?

Actually, it comes from NeoX dataset naming.  https://github.com/EleutherAI/gpt-neox/tree/main#preconfigured-datasets
In this test case, we use the following command to retrieve enwik8 (which is pile_subset) 
```
python prepare_data.py -d ./data -t HFTokenizer --vocab-file ./20B_checkpoints/20B_tokenizer.json pile_subset
```
but you can change it to full dataset retrieval by removing `_subset`.
```
python prepare_data.py -d ./data -t HFTokenizer --vocab-file ./20B_checkpoints/20B_tokenizer.json pile
```
The comment intended to direct those dataset into different directories.

File: 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+# Install GPT-NeoX  and its dependencies
+RUN git clone https://github.com/EleutherAI/gpt-neox.git \
+    && cd gpt-neox \
+    && pip install -r requirements/requirements.txt \
+    && pip install -r requirements/requirements-wandb.txt # optional, if logging using WandB \ 
+    && pip install -r requirements/requirements-tensorboard.txt # optional, if logging via tensorboard \
+    && python ./megatron/fused_kernels/setup.py install # optional, if using fused kernels 
+# Rebuild newer flash-attn
+RUN FLASH_ATTENTION_FORCE_BUILD=TRUE pip install flash-attn==2.5.5 --upgrade
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile:194]: You also need to set `MAX_JOBS` when building flash attention to control how much cpu cores and RAM used.

File: 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    # On p4de.24xlarge:
+    # - MAX_JOBS=16 => 145GB memory
+    # - MAX_JOBS=32 => 241GB memory
+    # - MAX_JOBS=48 => 243GB memory, 542.5s
+    #
+    # NOTE: must export MAX_JOBS. For some reason, `MAX_JOBS=16 pip install ...` doesn't seem to
+    #       work to prevent OOM.
+    export MAX_JOBS=32 && \
+    export NVCC_PREPEND_FLAGS="-t 32" && \
+    pip install -v -U git+https://github.com/facebookresearch/xformers.git@main#egg=xformers
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile]: Thanks Verdi for the comment. I have chosen the first path. 

In 8022077f8450994112cd22b79273b88ba1beda82 , I have added a script to let user to build container image on a compute instance and document the alternative option in `README.md`.

[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile]: and 1c25cf70578f6979e2038a9e3eb32a65f61aa0bf

File: 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile (hunk 8 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile]: Added variable in the dockerfdile.

File: 3.test_cases/15.gpt-neox/README.md (hunk 7 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/README.md:38]: Thanks for catching this. I have documented a way to tweak memory footprint and also provided alternative way to build container on a compute node.

File: 3.test_cases/15.gpt-neox/README.md (hunk 6 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/README.md:19]: Test case migrated to C4 dataset.

File: 3.test_cases/15.gpt-neox/2.train.sbatch (hunk 5 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/2.train.sbatch:23]: Thanks for catching this. I have addressed in https://github.com/aws-samples/awsome-distributed-training/pull/174/commits/7c03634d122c6bf5ee48e6f0a759a745d743d073

File: 3.test_cases/15.gpt-neox/1.prepare-data.sbatch (hunk 4 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/1.prepare-data.sbatch:20]: Test case migrated to C4 dataset.

File: 3.test_cases/15.gpt-neox/1.prepare-data.sbatch (hunk 3 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/1.prepare-data.sbatch:17]: Test case migrated to C4 dataset.

File: 3.test_cases/15.gpt-neox/1.prepare-data.sbatch (hunk 2 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/1.prepare-data.sbatch:16]: Test case migrated to C4 dataset.

**Comment by KeitaW** [comment]
https://github.com/aws-samples/awsome-distributed-training/pull/174/commits/b309e637fcb0818960e175ffea726fe2901842a6 Introduces changes needed to retrieve and process C4 dataset instead of the Pile dataset.

File: 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+#     # Load image to local docker registry -> on head node, or new compute/build node.
+#     docker load < /fsx/nvidia-pt-od__latest.tar
+####################################################################################################
+FROM nvcr.io/nvidia/pytorch:23.12-py3
+ENV DEBIAN_FRONTEND=noninteractive
+
+# The three must-be-built packages.
+# Efa-installer>=1.29.0 required for nccl>=2.19.0 to avoid libfabric NCCL error.
+ENV EFA_INSTALLER_VERSION=1.30.0
+ENV AWS_OFI_NCCL_VERSION=1.7.4-aws
```
[the remark this answers is not in the document]
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile:28]: Confirmed that works for the test case. Bumped

File: 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+#     docker load < /fsx/nvidia-pt-od__latest.tar
+####################################################################################################
+FROM nvcr.io/nvidia/pytorch:23.12-py3
+ENV DEBIAN_FRONTEND=noninteractive
+
+# The three must-be-built packages.
+# Efa-installer>=1.29.0 required for nccl>=2.19.0 to avoid libfabric NCCL error.
+ENV EFA_INSTALLER_VERSION=1.30.0
+ENV AWS_OFI_NCCL_VERSION=1.8.1-aws
+ENV NCCL_VERSION=2.19.3-1
```
Comment (verdimrc) [inline 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile:29]: Nits: parent container 23.12 is already on nccl-2.19.3 -- https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel-23-12.html#rel-23-12

**Review by verdimrc — APPROVED** [review]

File: 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile (hunk 11 above)
Comment (KeitaW) [reply 3.test_cases/15.gpt-neox/0.gpt-neox.dockerfile:29]: I have removed custom NCCL installation. Will test the container before merge this PR.
https://github.com/aws-samples/awsome-distributed-training/pull/174/commits/4f5ba3a432f76ef9671ce2f5fa37475dd1af66ea

**Comment by KeitaW** [comment]
Thanks @verdimrc for reviewing the PR! I will test the container again with pre-installed NCCL before merging this.

**Pull request merged**